newinterface.ts

```diff
@@ -0,0 +1,137 @@
+// See the end for stronger definitions of the above.
+
+// The first of the "smart" wrappers around BCD objects. Here, a browser object
+// contains some releases and queries for getting at them.
+declare class Browser {
+  id: browserKey;
+  data: object; // Underlying BCD browser object
+
+  // Pass through BCD values where convenient
+  name: string;
+  type: string;
+  preview_name: string;
+  // etc.
+
+  constructor(id: browserKey, data: unknown);
+
+  releaseAt(index: number): Release | undefined;
+  releaseOn(date: string): Release | undefined;
+  releasesFrom(index: number): Release[] | undefined;
+  releasesSince(date: string): Release[] | undefined;
+
+  // The last method here is mostly for convenience on the common case of "does
+  // this release support a given feature?". It's sugar for
+  // `browser(browserKey).releaseAt(0).supports(feature)`.
+  supports(feature: Feature): boolean | undefined;
+}
+
+// A `Release` represents concrete browser release. That is, it represents a
+// specific browser-version pair (i.e., it's unambiguously Chrome 100 or Firefox
+// 100). Releases don't imply a bunch of methods, but we'll be using these
+// objects in lots of places.
+declare class Release {
+  browser: Browser;
+  data: unknown; // Underlying BCD release object
+
+  // pass through BCD
+  version: string;
+  release_date: string;
+  status: string;
+  name: string; // We could even pass through the underlying browser data
+  // etc.
+
+  constructor(browser: Browser, releaseData: unknown);
+
+  // Mostly for symmetry with `Browser` and `Feature`. It's sugar for
+  // `feature.supportedBy(this)`;
+  supports(feature: Feature): boolean;
+}
+
+// This is where things get exciting. `Feature` is a wrapper around a `__compat`
+// object that's `Release`-aware.
+declare class Feature {
+  id: string; // dotted.path.to.feature
+  data: unknown; // underlying BCD `__compat` object
+
+  constructor(id: string, data: unknown);
+
+  // The simple case is to show a `Feature` a `Release` object and let it
+  // respond with boolean-ish value.
+  supportedBy(release: Release): supported;
+
+  // But there's no reason we can't pass multiple releases in (and `&&` them
+  // together, or near enough to it).
+  supportedBy(releases: Release[]): supported;
+  supportedBy(...releases: Release[]): supported;
+
+  // But BCD is very nuanced, so we'll need to offer some options, particularly around partial implementations.
+  supportedBy(releases: Release[], options: supportedByOptions): supported;
+
+  // Partials and notes will probably require some interrogation. Likewise, some
+  // consumers might want a more complex resolution to the question of multiple
+  // releases having potentially-conflicting supoprt values. For them, we offer
+  // the details for every release as a Map.
+  supportDetailsFor(
+    releases: Release[],
+    options: supportedByOptions,
+  ): Map<Release, { supported: supported; notes: Note }>;
```

Comment:
I wonder if we should have an object representing a support statement, and be able to return the support statement(s) which applies to a specific browser release? That will be an internal step needed for the `supportedBy` methods above anyway.

Reply:
Yes, I assumed that as an implementation detail we'd have a `SupportStatement` class somewhere, though I didn't give much thought to it as an API to expose. It's a bit fiddly to create one that stands alone, since you have to know the context (feature and browser) to make sense of it. I guess it'd look something like this:

```typescript
declare class SupportStatement {
  feature: Feature;
  browser: browserKey;
  data: unknown;

  constructor(feature: Feature, browser: browserKey, statement: unknown);

  // Pass through BCD
  version_added: string | boolean;
  version_removed: string | boolean;
  // etc.

  // Make some things friendlier?
  get version_removed(): string | boolean;
  get notes(): string[];
  get partial_implementation(): boolean;

  toReleases(): Release[];  // Convert a support statement into a sequence of Releases that correspond it
  has(release: Release): supported; // Check if a release is in the range represented by this support statement 
}
```

I'll come back to this tomorrow to actually add it in.